clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/DcosProviderUtils.groovy

```diff
@@ -0,0 +1,169 @@
+   * @return Non-null but possibly empty collection of CacheData.
+   */
+  static Collection<CacheData> resolveRelationshipData(Cache cacheView, CacheData source, String relationship, Closure<Boolean> relFilter) {
+    Collection<String> filteredRelationships = source?.relationships[relationship]?.findAll(relFilter)
+    filteredRelationships ? cacheView.getAll(relationship, filteredRelationships) : []
+  }
+
+  /**
+   * Similar to {@link #resolveRelationshipData}, except resolves the relationships from a given collection of source CacheData.
+   * @param cacheView The Cache from which CacheData for the relationships will be resolved.
+   * @param sources The sources from which the relationships will be resolved.
+   * @param relationship The relationship to resolve on the the source CacheData.
+   * @param relFilter An optional CacheFilter that can filter out relationships.
+   * @return Non-null but possibly empty collection of CacheData.
+   */
+  static Collection<CacheData> resolveRelationshipDataForCollection(Cache cacheView, Collection<CacheData> sources, String relationship, CacheFilter cacheFilter = null) {
+    Set<String> relationships = sources.findResults { it.relationships[relationship] ?: [] }.flatten()
+    relationships ? cacheView.getAll(relationship, relationships, cacheFilter) : []
+  }
+
+  static String combineAppStackDetail(String appName, String stack, String detail) {
+    NameValidation.notEmpty(appName, "appName")
+
+    // Use empty strings, not null references that output "null"
+    stack = stack != null ? stack : ""
+
+    if (detail != null && !detail.isEmpty()) {
+      return appName + "-" + stack + "-" + detail
+    }
+
+    if (!stack.isEmpty()) {
+      return appName + "-" + stack
+    }
+
+    return appName
+  }
+
+  static ImageDescription buildImageDescription(String image) {
+
+    if (!image || image.isEmpty()) {
+      return null
+    }
+
+    def sIndex = image.indexOf('/')
+    def result = new ImageDescription()
+
+    // No slash means we only provided a repository name & optional tag.
+    if (sIndex < 0) {
+      result.repository = image
+    } else {
+      def sPrefix = image.substring(0, sIndex)
+
+      // Check if the content before the slash is a registry (either localhost, or a URL)
+      if (sPrefix.startsWith('localhost') || sPrefix.contains('.')) {
+        result.registry = sPrefix
+
+        image = image.substring(sIndex + 1)
+      }
+    }
+
+    def cIndex = image.indexOf(':')
+
+    if (cIndex < 0) {
+      result.repository = image
+    } else {
+      result.tag = image.substring(cIndex + 1)
+      result.repository = image.subSequence(0, cIndex)
+    }
+
+    normalizeImageDescription(result)
+    result
+  }
+
+  static Void normalizeImageDescription(ImageDescription image) {
+    if (!image.registry) {
+      image.registry = "hub.docker.com" // TODO configure or pull from docker registry account
```

Comment:
Would this be less hidden as a default value of the `ImageDescription.registry` field itself?